### PR TITLE
osd: replace object_info_t::operator=() with decode()

### PR DESCRIPTION
**Pull request by jimmyway**
Signed-off-by: tang.jin <tang.jin@istuary.com>

### CODE REVIEW

**Comment by jimmyway** [comment]
@liupan1111

**Comment by jimmyway** [comment]
@dang

**Comment by jimmyway** [comment]
@danderson

File: src/osd/osd_types.h [hunk 1]
```diff
@@ -4117,7 +4117,7 @@ struct object_info_t {
   explicit object_info_t(bufferlist& bl) {
     decode(bl);
   }
-  object_info_t operator=(bufferlist& bl) {
+  object_info_t &operator=(bufferlist& bl) {
```
Comment (tchaikov) [inline src/osd/osd_types.h]: i think we should remove this method instead. and use `decode()` when appropriate. 

@dzafman what do you think?

Comment (jimmyway) [inline src/osd/osd_types.h]: @tchaikov  May I delete this function directly or append  '= delete' to the function declaration?

Comment (tchaikov) [inline src/osd/osd_types.h]: as long as there is not users of `operator=`, i don't see why we cannot delete the default copy assignment operator.

Comment (jimmyway) [inline src/osd/osd_types.h]: OK, I will delete it.

Comment (tchaikov) [inline src/osd/osd_types.h]: @jimmyway please try compile the tree first. `PG::_scan_snaps()` uses this function, that's why i suggested use `decode()` directly.

Comment (tchaikov) [inline src/osd/osd_types.h]: and what is deleted in your change is not the default copy assignment operator. the one created by compiler by default is 
```c++
object_info_t& operator=(object_info_t&)
```

Comment (jimmyway) [inline src/osd/osd_types.h]: sorry, I made mistakes, I will check it at first.

Comment (jimmyway) [inline src/osd/osd_types.h]: Do you mean delete 'operation=' member function, and replace it to decode() in PG::_scan_snaps() ?

**Comment by jimmyway** [comment]
@tchaikov

**Comment by yuriw** [comment]
http://pulpito.ceph.com/yuriw-2017-03-18_00:03:11-rados-wip-yuri-testing2_2017_3_18-distro-basic-smithi/

http://pulpito.ceph.com/yuriw-2017-03-18_15:53:35-rados-wip-yuri-testing2_2017_3_18---basic-smithi/

100% tests passed, 0 tests failed out of 163

**Pull request merged**